In [1]:
import os
import pandas as pd
import numpy as np
pd.set_option("display.max_colwidth", 2)

os.chdir("/home/coinse/Workspace/SBFL")

# Load bug info

In [2]:
from scripts.util import load_bug_data, find_covmat_func, find_covmat_idx
from functools import reduce
from typing import Set

pid, vid = "wireshark", 6

bug = load_bug_data(pid, vid)

print(f"{bug.buggy_files=}")
print(f"{bug.buggy_lines=}")
print("bug.coverage_df=")
display(bug.coverage_df)
print(f"{bug.failing_tests=}")
print(f"{bug.failing_tests_GT=}")
print("bug.functions={", end="")
for func in list(bug.functions)[:5]:
    print(f"{func},", end=" ")
print(", ...}")
print(f"{bug.id=}")
print(f"{bug.num_test_cases=}")
print(f"{bug.passing_tests=}")
print(f"{bug.path_to_cov_df=}")
print(f"{bug.path_to_patch=}")
print(f"{bug.path_to_test_results=}")
print("bug.source_files=", end="")
for src in list(bug.source_files)[:5]:
    print(f"{src},", end=" ")
print(", ...}")

fault_idxs = sorted(bug.buggy_lines)
# fault_idx2meth = {}
fault_meth2idx = {}
err_msg = ""
for fault_idx in fault_idxs:
    succ_func, ret_func = find_covmat_func(
        pid, bug.coverage_df, fault_idx, debug=False
    )
    if succ_func:
        succ_idx, ret_idx = find_covmat_idx(
            pid, bug.coverage_df, fault_idx, debug=False
        )
        if succ_idx:
            if ret_func in fault_meth2idx:
                fault_meth2idx[ret_func].add(ret_idx)
            else:
                fault_meth2idx[ret_func] = {ret_idx}
        else:
            print(f"{ret_idx}")
            if ret_func not in fault_meth2idx:
                fault_meth2idx[ret_func] = set()
    else:
        print(f"{ret_func}")
# fault_meths: Set = reduce(
#     lambda x, y: x | set([y]), fault_idx2meth.values(), set()
# )
# fault_meth2idx = {
#     v: {idx for idx in fault_idx2meth if fault_idx2meth[idx] == v}
#     for v in fault_meths
# }
fault_meths = set(fault_meth2idx.keys())
for meth in fault_meth2idx:
    print(f"{meth=}: {fault_meth2idx[meth]}")
# print(f"{fault_meths=}")


bug.buggy_files={'epan/wslua/wslua_byte_array.c'}
bug.buggy_lines={('epan/wslua/wslua_byte_array.c', 267), ('epan/wslua/wslua_byte_array.c', 266), ('epan/wslua/wslua_byte_array.c', 262), ('epan/wslua/wslua_byte_array.c', 268), ('epan/wslua/wslua_byte_array.c', 265), ('epan/wslua/wslua_byte_array.c', 261)}
bug.coverage_df=


33  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   556   
                                                                                  177   556   
                                                                                  199   556   
                                                                                  203   556   
                                                                                  205   556   
...                                                                                     ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1712  556   
                                                                                  1713  556   
                                                                                  1715  556   
                                                                                  1716  556   
                                                                                  1717  556   

                                                                                         32  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   554   
                                                                                  177   554   
                                                                                  199   554   
                                                                                  203   554   
                                                                                  205   554   
...                                                                                     ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1712  554   
                                                                                  1713  554   
                                                                                  1715  554   
                                                                                  1716  554   
                                                                                  1717  554   

                                                                                          9  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   278   
                                                                                  177   278   
                                                                                  199   278   
                                                                                  203   278   
                                                                                  205   278   
...                                                                                     ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1712  278   
                                                                                  1713  278   
                                                                                  1715  278   
                                                                                  1716  278   
                                                                                  1717  278   

                                                                                         10  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   280   
                                                                                  177   280   
 

bug.failing_tests={'34'}
bug.failing_tests_GT={'34'}
bug.functions={('/home/workspace/epan/dissectors/packet-tls-utils.c', 'ssl3_check_mac'), ('/home/workspace/epan/dissectors/packet-usbip.c', 'proto_reg_handoff_usbip'), ('/home/workspace/epan/wmem/wmem_strutl.c', 'wmem_strdup_printf'), ('/home/workspace/epan/proto.c', 'proto_tree_add_double_format'), ('/home/workspace/wiretap/file_wrappers.c', 'file_clearerr'), , ...}
bug.id='wireshark-6'
bug.num_test_cases=34
bug.passing_tests={'3', '15', '29', '16', '11', '2', '27', '13', '22', '26', '17', '25', '32', '23', '24', '12', '4', '8', '1', '19', '18', '30', '14', '33', '28', '7', '9', '31', '20', '6', '5', '10', '21'}
bug.path_to_cov_df='/home/coinse/Workspace/defects4cpp/coverage/wireshark_6_buggy/coverage_df.pkl'
bug.path_to_patch='/home/coinse/Workspace/defects4cpp/defects4cpp/taxonomy/wireshark/patch/0006-buggy.patch'
bug.path_to_test_results='/home/coinse/Workspace/defects4cpp/coverage/wireshark_6_buggy/test_results.json'
bug.source_

# Get spectra

In [3]:
covmat = bug.coverage_df
pass_tcs = bug.passing_tests
fail_tcs = bug.failing_tests

covmat_bool = covmat.astype(bool)
P = len(pass_tcs)
F = len(fail_tcs)
e_p = covmat_bool[pass_tcs].sum(axis=1)
e_f = covmat_bool[fail_tcs].sum(axis=1)
n_p = P - e_p
n_f = F - e_f
spectra = pd.DataFrame(
    {
        "line": covmat.index.to_list(),
        "e_p": e_p,
        "e_f": e_f,
        "n_p": n_p,
        "n_f": n_f,
    }
)

display(spectra)


line  \
file                                                  function                    line                                                                                              
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 174)   
                                                                                  177   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 177)   
                                                                                  199   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 199)   
                                                                                  203   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 203)   
                                                                                  205   (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap, 205)   
...                                                                                                                                                                           ...   
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1712  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1712)               
                                                                                  1713  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1713)               
                                                                                  1715  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1715)               
                                                                                  1716  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1716)               
                                                                                  1717  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns, 1717)               

                                                                                        e_p  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   33    
                                                                                  177   33    
                                                                                  199   33    
                                                                                  203   33    
                                                                                  205   33    
...                                                                                     ..    
/home/workspace/epan/dissectors/packet-isns.c         proto_reg_handoff_isns      1712  33    
                                                                                  1713  33    
                                                                                  1715  33    
                                                                                  1716  33    
                                                                                  1717  33    

                                                                                        e_f  \
file                                                  function                    line        
/home/workspace/epan/dissectors/packet-gluster_pmap.c proto_register_gluster_pmap 174   1     
                                                                                  177   1     
                                                                                  199   1     
                                                                                  203   1     
             

# Compute suspiciousness scores

In [4]:
def ochiai(e_f, e_p, n_f, n_p):
    return e_f / np.sqrt(e_p + e_f) / np.sqrt(e_f + n_f) if e_f else 0.0


def op2(e_f, e_p, n_f, n_p):
    return e_f - e_p / (e_p + n_p + 1)

spectra["Sop2"] = spectra.apply(
    lambda row: op2(row.e_f, row.e_p, row.n_f, row.n_p), axis=1
)
spectra["Sochiai"] = spectra.apply(
    lambda row: ochiai(row.e_f, row.e_p, row.n_f, row.n_p), axis=1
)
spectra["Rop2"] = spectra.Sop2.rank(ascending=False, method="max")
spectra["Rochiai"] = spectra.Sochiai.rank(ascending=False, method="max")


# Line level SBFL

In [5]:
for fault_meth in fault_meth2idx:
    fault_idxs_of_meth = fault_meth2idx[fault_meth]
    print("fault meth:", fault_meth)
    for fault_idx in fault_idxs_of_meth:
        print("fault idx:", tuple(fault_idx))
    print("SBFL score of faulty method:")
    display(spectra.loc[fault_idxs_of_meth].reset_index(drop=True))

print("Top 5 SBFL scores:")
print("Op2:")
display(
    spectra.sort_values("Sop2", ascending=False).head(5).reset_index(drop=True)
)
print("Ochiai:")
display(
    spectra.sort_values("Sochiai", ascending=False)
    .head(5)
    .reset_index(drop=True)
)


fault meth: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode')
fault idx: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode', 267)
fault idx: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode', 266)
fault idx: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode', 265)
fault idx: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode', 262)
fault idx: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode', 268)
fault idx: ('/home/workspace/epan/wslua/wslua_byte_array.c', 'ByteArray_base64_decode', 261)
SBFL score of faulty method:


line  \
0  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode, 267)   
1  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode, 266)   
2  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode, 265)   
3  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode, 262)   
4  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode, 268)   
5  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode, 261)   

   e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  
0  0    1    33   0    1.0   1.0      4981.0  4981.0   
1  0    1    33   0    1.0   1.0      4981.0  4981.0   
2  0    1    33   0    1.0   1.0      4981.0  4981.0   
3  0    1    33   0    1.0   1.0      4981.0  4981.0   
4  0    1    33   0    1.0   1.0      4981.0  4981.0   
5  0    1    33   0    1.0   1.0      4981.0  4981.0

Top 5 SBFL scores:
Op2:


line  e_p  \
0  (/home/workspace/wiretap/k12.c, k12_open, 823)                      0     
1  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_lo, 382)       0     
2  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_private, 335)  0     
3  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_private, 334)  0     
4  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_columns, 330)  0     

   e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  
0  1    33   0    1.0   1.0      4981.0  4981.0   
1  1    33   0    1.0   1.0      4981.0  4981.0   
2  1    33   0    1.0   1.0      4981.0  4981.0   
3  1    33   0    1.0   1.0      4981.0  4981.0   
4  1    33   0    1.0   1.0      4981.0  4981.0

Ochiai:


line  e_p  \
0  (/home/workspace/wiretap/k12.c, k12_open, 823)                      0     
1  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_lo, 382)       0     
2  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_private, 335)  0     
3  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_private, 334)  0     
4  (/home/workspace/epan/wslua/wslua_pinfo.c, Pinfo_get_columns, 330)  0     

   e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  
0  1    33   0    1.0   1.0      4981.0  4981.0   
1  1    33   0    1.0   1.0      4981.0  4981.0   
2  1    33   0    1.0   1.0      4981.0  4981.0   
3  1    33   0    1.0   1.0      4981.0  4981.0   
4  1    33   0    1.0   1.0      4981.0  4981.0

# Method level SBFL

In [6]:
spectra["file"] = spectra.index.map(lambda x: x[0])
spectra["func"] = spectra.index.map(lambda x: x[1])
spectra["line"] = spectra.index.map(lambda x: x[2])
# (file, func) is meth
spectra["meth"] = spectra.apply(lambda row: (row.file, row.func), axis=1)
spectra.reset_index(drop=True, inplace=True)
display(spectra)

line  e_p  e_f  n_p  n_f      Sop2   Sochiai     Rop2  Rochiai  \
0      174   33   1    0    0    0.029412  0.171499  98416.0  98416.0   
1      177   33   1    0    0    0.029412  0.171499  98416.0  98416.0   
2      199   33   1    0    0    0.029412  0.171499  98416.0  98416.0   
3      203   33   1    0    0    0.029412  0.171499  98416.0  98416.0   
4      205   33   1    0    0    0.029412  0.171499  98416.0  98416.0   
...    ...   ..  ..   ..   ..         ...       ...      ...      ...   
98537  1712  33   1    0    0    0.029412  0.171499  98416.0  98416.0   
98538  1713  33   1    0    0    0.029412  0.171499  98416.0  98416.0   
98539  1715  33   1    0    0    0.029412  0.171499  98416.0  98416.0   
98540  1716  33   1    0    0    0.029412  0.171499  98416.0  98416.0   
98541  1717  33   1    0    0    0.029412  0.171499  98416.0  98416.0   

                                                        file  \
0      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
1      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
2      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
3      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
4      /home/workspace/epan/dissectors/packet-gluster_pmap.c   
...                                                      ...   
98537  /home/workspace/epan/dissectors/packet-isns.c           
98538  /home/workspace/epan/dissectors/packet-isns.c           
98539  /home/workspace/epan/dissectors/packet-isns.c           
98540  /home/workspace/epan/dissectors/packet-isns.c           
98541  /home/workspace/epan/dissectors/packet-isns.c           

                              func  \
0      proto_register_gluster_pmap   
1      proto_register_gluster_pmap   
2      proto_register_gluster_pmap   
3      proto_register_gluster_pmap   
4      proto_register_gluster_pmap   
...                            ...   
98537  proto_reg_handoff_isns        
98538  proto_reg_handoff_isns        
98539  proto_reg_handoff_isns        
98540  proto_reg_handoff_isns        
98541  proto_reg_handoff_isns        

                                                                                       meth  
0      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
1      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
2      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
3      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
4      (/home/workspace/epan/dissectors/packet-gluster_pmap.c, proto_register_gluster_pmap)  
...                                                                                     ...  
98537  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
98538  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
98539  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
98540  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               
98541  (/home/workspace/epan/dissectors/packet-isns.c, proto_reg_handoff_isns)               

[98542 rows x 12 columns]

## Max aggregation

In [7]:
print("[Op2]")
spectra_max_op2 = spectra.copy().sort_values(
    ["Sop2", "meth", "line"], ascending=False
)
spectra_max_op2.drop_duplicates(subset="meth", keep="first", inplace=True)
spectra_max_op2 = spectra_max_op2[
    ["meth", "line", "e_p", "e_f", "n_p", "n_f", "Sop2"]
]
spectra_max_op2["Rop2"] = spectra_max_op2.Sop2.rank(
    ascending=False, method="max"
)
print("Faulty method:")
display(
    spectra_max_op2.loc[spectra_max_op2.meth.isin(fault_meths)].reset_index(
        drop=True
    )
)
print("Top 20 methods:")
display(spectra_max_op2.head(20).reset_index(drop=True))
print()
print("[Ochiai]")
spectra_max_ochiai = spectra.copy().sort_values(
    ["Sochiai", "meth", "line"], ascending=False
)
spectra_max_ochiai.drop_duplicates(subset="meth", keep="first", inplace=True)
spectra_max_ochiai = spectra_max_ochiai[
    ["meth", "line", "e_p", "e_f", "n_p", "n_f", "Sochiai"]
]
spectra_max_ochiai["Rochiai"] = spectra_max_ochiai.Sochiai.rank(
    ascending=False, method="max"
)
print("Faulty method:")
display(
    spectra_max_ochiai.loc[
        spectra_max_ochiai.meth.isin(fault_meths)
    ].reset_index(drop=True)
)
print("Top 20 methods:")
display(spectra_max_ochiai.head(20).reset_index(drop=True))


[Op2]
Faulty method:


meth  \
0  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   

   line  e_p  e_f  n_p  n_f  Sop2   Rop2  
0  276   0    1    33   0    1.0   683.0

Top 20 methods:


meth  \
0   (/home/workspace/wsutil/report_message.c, report_failure)                          
1   (/home/workspace/wsutil/nstime.c, nstime_sum)                                      
2   (/home/workspace/wsutil/filesystem.c, get_dirname)                                 
3   (/home/workspace/wiretap/wtap.c, wtap_sequential_close)                            
4   (/home/workspace/wiretap/visual.c, visual_open)                                    
5   (/home/workspace/wiretap/tnef.c, tnef_open)                                        
6   (/home/workspace/wiretap/snoop.c, snoop_open)                                      
7   (/home/workspace/wiretap/rfc7468.c, rfc7468_open)                                  
8   (/home/workspace/wiretap/rfc7468.c, read_complete_text_line)                       
9   (/home/workspace/wiretap/radcom.c, radcom_open)                                    
10  (/home/workspace/wiretap/peektagged.c, peektagged_open)                            
11  (/home/workspace/wiretap/pcapng.c, pcapng_write_enhanced_packet_block)             
12  (/home/workspace/wiretap/pcapng.c, pcapng_read_block)                              
13  (/home/workspace/wiretap/pcapng.c, pcapng_open)                                    
14  (/home/workspace/wiretap/ngsniffer.c, ngsniffer_open)                              
15  (/home/workspace/wiretap/netxray.c, netxray_open)                                  
16  (/home/workspace/wiretap/network_instruments.c, network_instruments_open)          
17  (/home/workspace/wiretap/nettrace_3gpp_32_423.c, nettrace_3gpp_32_423_file_open)   
18  (/home/workspace/wiretap/nettl.c, nettl_open)                                      
19  (/home/workspace/wiretap/netmon.c, netmon_open)                                    

    line  e_p  e_f  n_p  n_f  Sop2   Rop2  
0   54    0    1    33   0    1.0   683.0  
1   127   0    1    33   0    1.0   683.0  
2   154   0    1    33   0    1.0   683.0  
3   1319  0    1    33   0    1.0   683.0  
4   176   0    1    33   0    1.0   683.0  
5   24    0    1    33   0    1.0   683.0  
6   249   0    1    33   0    1.0   683.0  
7   80    0    1    33   0    1.0   683.0  
8   41    0    1    33   0    1.0   683.0  
9   103   0    1    33   0    1.0   683.0  
10  268   0    1    33   0    1.0   683.0  
11  3736  0    1    33   0    1.0   683.0  
12  2697  0    1    33   0    1.0   683.0  
13  2871  0    1    33   0    1.0   683.0  
14  572   0    1    33   0    1.0   683.0  
15  460   0    1    33   0    1.0   683.0  
16  144   0    1    33   0    1.0   683.0  
17  1186  0    1    33   0    1.0   683.0  
18  193   0    1    33   0    1.0   683.0  
19  432   0    1    33   0    1.0   683.0


[Ochiai]
Faulty method:


meth  \
0  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   

   line  e_p  e_f  n_p  n_f  Sochiai  Rochiai  
0  276   0    1    33   0    1.0      683.0

Top 20 methods:


meth  \
0   (/home/workspace/wsutil/report_message.c, report_failure)                          
1   (/home/workspace/wsutil/nstime.c, nstime_sum)                                      
2   (/home/workspace/wsutil/filesystem.c, get_dirname)                                 
3   (/home/workspace/wiretap/wtap.c, wtap_sequential_close)                            
4   (/home/workspace/wiretap/visual.c, visual_open)                                    
5   (/home/workspace/wiretap/tnef.c, tnef_open)                                        
6   (/home/workspace/wiretap/snoop.c, snoop_open)                                      
7   (/home/workspace/wiretap/rfc7468.c, rfc7468_open)                                  
8   (/home/workspace/wiretap/rfc7468.c, read_complete_text_line)                       
9   (/home/workspace/wiretap/radcom.c, radcom_open)                                    
10  (/home/workspace/wiretap/peektagged.c, peektagged_open)                            
11  (/home/workspace/wiretap/pcapng.c, pcapng_write_enhanced_packet_block)             
12  (/home/workspace/wiretap/pcapng.c, pcapng_read_block)                              
13  (/home/workspace/wiretap/pcapng.c, pcapng_open)                                    
14  (/home/workspace/wiretap/ngsniffer.c, ngsniffer_open)                              
15  (/home/workspace/wiretap/netxray.c, netxray_open)                                  
16  (/home/workspace/wiretap/network_instruments.c, network_instruments_open)          
17  (/home/workspace/wiretap/nettrace_3gpp_32_423.c, nettrace_3gpp_32_423_file_open)   
18  (/home/workspace/wiretap/nettl.c, nettl_open)                                      
19  (/home/workspace/wiretap/netmon.c, netmon_open)                                    

    line  e_p  e_f  n_p  n_f  Sochiai  Rochiai  
0   54    0    1    33   0    1.0      683.0    
1   127   0    1    33   0    1.0      683.0    
2   154   0    1    33   0    1.0      683.0    
3   1319  0    1    33   0    1.0      683.0    
4   176   0    1    33   0    1.0      683.0    
5   24    0    1    33   0    1.0      683.0    
6   249   0    1    33   0    1.0      683.0    
7   80    0    1    33   0    1.0      683.0    
8   41    0    1    33   0    1.0      683.0    
9   103   0    1    33   0    1.0      683.0    
10  268   0    1    33   0    1.0      683.0    
11  3736  0    1    33   0    1.0      683.0    
12  2697  0    1    33   0    1.0      683.0    
13  2871  0    1    33   0    1.0      683.0    
14  572   0    1    33   0    1.0      683.0    
15  460   0    1    33   0    1.0      683.0    
16  144   0    1    33   0    1.0      683.0    
17  1186  0    1    33   0    1.0      683.0    
18  193   0    1    33   0    1.0      683.0    
19  432   0    1    33   0    1.0      683.0

## Line2Method voting (weight: sbfl score / dense rank)

In [8]:
print("[Op2]")
ballot_op2 = {}
spectra_max_op2 = spectra.copy().sort_values(
    ["Sop2", "meth", "line"], ascending=False
)
spectra_max_op2["Rop2D"] = spectra_max_op2.Sop2.rank(
    ascending=False, method="dense"
)
spectra_max_op2["vote"] = spectra_max_op2.Sop2 / spectra_max_op2.Rop2D
for row in spectra_max_op2.itertuples():
    if row.meth not in ballot_op2:
        ballot_op2[row.meth] = row.vote
    else:
        ballot_op2[row.meth] += row.vote
L2M_op2 = pd.DataFrame(
    {"meth": list(ballot_op2.keys()), "vote": list(ballot_op2.values())}
)
L2M_op2["rank"] = L2M_op2.vote.rank(ascending=False, method="max")
L2M_op2.sort_values("rank", inplace=True)
L2M_op2.reset_index(drop=True, inplace=True)
for fault_meth in fault_meths:
    print("Faulty method:")
    display(L2M_op2.loc[L2M_op2.meth == fault_meth].reset_index(drop=True))
    print("Voters:")
    display(
        spectra_max_op2.loc[spectra_max_op2.meth == fault_meth].reset_index(
            drop=True
        )
    )
print("Top 5 methods:")
display(L2M_op2.head(5).reset_index(drop=True))
for idx, top_meth in enumerate(L2M_op2.head(5).meth.tolist(), 1):
    print(f"Top {idx}:")
    display(L2M_op2.loc[L2M_op2.meth == top_meth].reset_index(drop=True))
    print("Voters:")
    display(
        spectra_max_op2.loc[spectra_max_op2.meth == top_meth].reset_index(
            drop=True
        )
    )
print()
print("[Ochiai]")
ballot_ochiai = {}
spectra_max_ochiai = spectra.copy().sort_values(
    ["Sochiai", "meth", "line"], ascending=False
)
spectra_max_ochiai["RochiaiD"] = spectra_max_ochiai.Sochiai.rank(
    ascending=False, method="dense"
)
spectra_max_ochiai["vote"] = (
    spectra_max_ochiai.Sochiai / spectra_max_ochiai.RochiaiD
)
for row in spectra_max_ochiai.itertuples():
    if row.meth not in ballot_ochiai:
        ballot_ochiai[row.meth] = row.vote
    else:
        ballot_ochiai[row.meth] += row.vote
L2M_ochiai = pd.DataFrame(
    {"meth": list(ballot_ochiai.keys()), "vote": list(ballot_ochiai.values())}
)
L2M_ochiai["rank"] = L2M_ochiai.vote.rank(ascending=False, method="max")
L2M_ochiai.sort_values("rank", inplace=True)
L2M_ochiai.reset_index(drop=True, inplace=True)
for fault_meth in fault_meths:
    print("Faulty method:")
    display(
        L2M_ochiai.loc[L2M_ochiai.meth == fault_meth].reset_index(drop=True)
    )
    print("Voters:")
    display(
        spectra_max_ochiai.loc[
            spectra_max_ochiai.meth == fault_meth
        ].reset_index(drop=True)
    )
print("Top 5 methods:")
display(L2M_ochiai.head(5).reset_index(drop=True))
for idx, top_meth in enumerate(L2M_ochiai.head(5).meth.tolist(), 1):
    print(f"Top {idx}:")
    display(L2M_ochiai.loc[L2M_ochiai.meth == top_meth].reset_index(drop=True))
    print("Voters:")
    display(
        spectra_max_ochiai.loc[spectra_max_ochiai.meth == top_meth].reset_index(
            drop=True
        )
    )


[Op2]
Faulty method:


meth  \
0  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   

   vote   rank  
0  15.0  128.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   276   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   275   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   272   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   271   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   270   0    1    33   0    1.0   1.0      4981.0  4981.0    
5   268   0    1    33   0    1.0   1.0      4981.0  4981.0    
6   267   0    1    33   0    1.0   1.0      4981.0  4981.0    
7   266   0    1    33   0    1.0   1.0      4981.0  4981.0    
8   265   0    1    33   0    1.0   1.0      4981.0  4981.0    
9   264   0    1    33   0    1.0   1.0      4981.0  4981.0    
10  262   0    1    33   0    1.0   1.0      4981.0  4981.0    
11  261   0    1    33   0    1.0   1.0      4981.0  4981.0    
12  260   0    1    33   0    1.0   1.0      4981.0  4981.0    
13  259   0    1    33   0    1.0   1.0      4981.0  4981.0    
14  254   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                             file                     func  \
0   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
1   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
2   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
3   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
4   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
5   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
6   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
7   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
8   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
9   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
10  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
11  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
12  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
13  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
14  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   

                                                                        meth  \
0   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
1   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
2   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
3   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
4   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
5   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
6   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
7   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
8   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
9   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
10  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
11  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
12  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
13  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
14  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   

    Rop2D  vote  
0   1.0    1.0   
1   1.0    1.0   
2   1.0    1.0   
3   1.0    1.0   
4   1.0    1.0   
5   1.0    1.0   
6   1.0    1.0   
7   1.0    1.0   
8   1.0    1.0   
9   1.0    1.0   
10  1.0    1.0   
11  1.0    1.0   
12  1.0    1.0   
13  1.0    1.0   
14  1.0    1.0

Top 5 methods:


meth  \
0  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)               
1  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)                     
2  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
3  (/home/workspace/epan/tvbtest.c, run_tests)                                          
4  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)                 

         vote  rank  
0  105.000000  1.0   
1  93.000000   2.0   
2  87.838235   3.0   
3  77.161765   4.0   
4  72.000000   5.0

Top 1:


meth  \
0  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   

    vote  rank  
0  105.0  1.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0    2887  0    1    33   0    1.0   1.0      4981.0  4981.0    
1    2886  0    1    33   0    1.0   1.0      4981.0  4981.0    
2    2884  0    1    33   0    1.0   1.0      4981.0  4981.0    
3    2881  0    1    33   0    1.0   1.0      4981.0  4981.0    
4    2877  0    1    33   0    1.0   1.0      4981.0  4981.0    
..    ... ..   ..    ..  ..    ...   ...         ...     ...    
100  2582  0    1    33   0    1.0   1.0      4981.0  4981.0    
101  2581  0    1    33   0    1.0   1.0      4981.0  4981.0    
102  2580  0    1    33   0    1.0   1.0      4981.0  4981.0    
103  2579  0    1    33   0    1.0   1.0      4981.0  4981.0    
104  2577  0    1    33   0    1.0   1.0      4981.0  4981.0    

                                             file                    func  \
0    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
1    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
2    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
3    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
4    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
..                                            ...                     ...   
100  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
101  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
102  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
103  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
104  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   

                                                                       meth  \
0    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
1    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
2    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
3    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
4    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
..                                                                      ...   
100  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
101  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
102  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
103  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
104  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   

     Rop2D  vote  
0    1.0    1.0   
1    1.0    1.0   
2    1.0    1.0   
3    1.0    1.0   
4    1.0    1.0   
..   ...    ...   
100  1.0    1.0   
101  1.0    1.0   
102  1.0    1.0   
103  1.0    1.0   
104  1.0    1.0   

[105 rows x 14 columns]

Top 2:


meth  vote  \
0  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  93.0   

   rank  
0  2.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   802   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   800   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   797   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   795   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   794   0    1    33   0    1.0   1.0      4981.0  4981.0    
..  ...  ..   ..    ..  ..    ...   ...         ...     ...    
88  538   0    1    33   0    1.0   1.0      4981.0  4981.0    
89  537   0    1    33   0    1.0   1.0      4981.0  4981.0    
90  536   0    1    33   0    1.0   1.0      4981.0  4981.0    
91  535   0    1    33   0    1.0   1.0      4981.0  4981.0    
92  508   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                              file            func  \
0   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
1   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
2   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
3   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
4   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
..                                             ...             ...   
88  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
89  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
90  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
91  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
92  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   

                                                                meth  Rop2D  \
0   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
1   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
2   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
3   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
4   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
..                                                               ...  ...     
88  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
89  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
90  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
91  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     
92  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  1.0     

    vote  
0   1.0   
1   1.0   
2   1.0   
3   1.0   
4   1.0   
..  ...   
88  1.0   
89  1.0   
90  1.0   
91  1.0   
92  1.0   

[93 rows x 14 columns]

Top 3:


meth  \
0  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   

        vote  rank  
0  87.838235  3.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai    Rop2  Rochiai  \
0    571   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
1    570   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
2    569   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
3    568   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
4    567   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
..   ...  ..   ..    ..  ..         ...       ...     ...     ...    
176  335   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
177  330   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
178  328   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
179  327   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
180  325   1    1    32   0    0.970588  0.707107  7651.0  7651.0    

                                       file  \
0    /home/workspace/epan/reassemble_test.c   
1    /home/workspace/epan/reassemble_test.c   
2    /home/workspace/epan/reassemble_test.c   
3    /home/workspace/epan/reassemble_test.c   
4    /home/workspace/epan/reassemble_test.c   
..                                      ...   
176  /home/workspace/epan/reassemble_test.c   
177  /home/workspace/epan/reassemble_test.c   
178  /home/workspace/epan/reassemble_test.c   
179  /home/workspace/epan/reassemble_test.c   
180  /home/workspace/epan/reassemble_test.c   

                                         func  \
0    test_fragment_add_seq_partial_reassembly   
1    test_fragment_add_seq_partial_reassembly   
2    test_fragment_add_seq_partial_reassembly   
3    test_fragment_add_seq_partial_reassembly   
4    test_fragment_add_seq_partial_reassembly   
..                                        ...   
176  test_fragment_add_seq_partial_reassembly   
177  test_fragment_add_seq_partial_reassembly   
178  test_fragment_add_seq_partial_reassembly   
179  test_fragment_add_seq_partial_reassembly   
180  test_fragment_add_seq_partial_reassembly   

                                                                                   meth  \
0    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
1    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
2    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
3    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
4    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
..                                                                                  ...   
176  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
177  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
178  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
179  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
180  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   

     Rop2D      vote  
0    2.0    0.485294  
1    2.0    0.485294  
2    2.0    0.485294  
3    2.0    0.485294  
4    2.0    0.485294  
..   ...         ...  
176  2.0    0.485294  
177  2.0    0.485294  
178  2.0    0.485294  
179  2.0    0.485294  
180  2.0    0.485294  

[181 rows x 14 columns]

Top 4:


meth       vote  rank
0  (/home/workspace/epan/tvbtest.c, run_tests)  77.161765  4.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai    Rop2  Rochiai  \
0    486   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
1    485   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
2    483   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
3    482   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
4    481   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
..   ...  ..   ..    ..  ..         ...       ...     ...     ...    
154  284   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
155  283   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
156  282   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
157  280   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
158  278   1    1    32   0    0.970588  0.707107  7651.0  7651.0    

                               file       func  \
0    /home/workspace/epan/tvbtest.c  run_tests   
1    /home/workspace/epan/tvbtest.c  run_tests   
2    /home/workspace/epan/tvbtest.c  run_tests   
3    /home/workspace/epan/tvbtest.c  run_tests   
4    /home/workspace/epan/tvbtest.c  run_tests   
..                              ...        ...   
154  /home/workspace/epan/tvbtest.c  run_tests   
155  /home/workspace/epan/tvbtest.c  run_tests   
156  /home/workspace/epan/tvbtest.c  run_tests   
157  /home/workspace/epan/tvbtest.c  run_tests   
158  /home/workspace/epan/tvbtest.c  run_tests   

                                            meth  Rop2D      vote  
0    (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
1    (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
2    (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
3    (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
4    (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
..                                           ...  ...         ...  
154  (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
155  (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
156  (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
157  (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  
158  (/home/workspace/epan/tvbtest.c, run_tests)  2.0    0.485294  

[159 rows x 14 columns]

Top 5:


meth  vote  \
0  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)  72.0   

   rank  
0  5.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   938   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   936   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   933   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   931   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   930   0    1    33   0    1.0   1.0      4981.0  4981.0    
..  ...  ..   ..    ..  ..    ...   ...         ...     ...    
67  809   0    1    33   0    1.0   1.0      4981.0  4981.0    
68  808   0    1    33   0    1.0   1.0      4981.0  4981.0    
69  807   0    1    33   0    1.0   1.0      4981.0  4981.0    
70  806   0    1    33   0    1.0   1.0      4981.0  4981.0    
71  805   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                              file                func  \
0   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
1   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
2   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
3   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
4   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
..                                             ...                 ...   
67  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
68  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
69  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
70  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
71  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   

                                                                    meth  \
0   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
1   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
2   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
3   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
4   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
..                                                                   ...   
67  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
68  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
69  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
70  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
71  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   

    Rop2D  vote  
0   1.0    1.0   
1   1.0    1.0   
2   1.0    1.0   
3   1.0    1.0   
4   1.0    1.0   
..  ...    ...   
67  1.0    1.0   
68  1.0    1.0   
69  1.0    1.0   
70  1.0    1.0   
71  1.0    1.0   

[72 rows x 14 columns]


[Ochiai]
Faulty method:


meth  \
0  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   

   vote   rank  
0  15.0  108.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   276   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   275   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   272   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   271   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   270   0    1    33   0    1.0   1.0      4981.0  4981.0    
5   268   0    1    33   0    1.0   1.0      4981.0  4981.0    
6   267   0    1    33   0    1.0   1.0      4981.0  4981.0    
7   266   0    1    33   0    1.0   1.0      4981.0  4981.0    
8   265   0    1    33   0    1.0   1.0      4981.0  4981.0    
9   264   0    1    33   0    1.0   1.0      4981.0  4981.0    
10  262   0    1    33   0    1.0   1.0      4981.0  4981.0    
11  261   0    1    33   0    1.0   1.0      4981.0  4981.0    
12  260   0    1    33   0    1.0   1.0      4981.0  4981.0    
13  259   0    1    33   0    1.0   1.0      4981.0  4981.0    
14  254   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                             file                     func  \
0   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
1   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
2   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
3   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
4   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
5   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
6   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
7   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
8   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
9   /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
10  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
11  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
12  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
13  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   
14  /home/workspace/epan/wslua/wslua_byte_array.c  ByteArray_base64_decode   

                                                                        meth  \
0   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
1   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
2   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
3   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
4   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
5   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
6   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
7   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
8   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
9   (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
10  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
11  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
12  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
13  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   
14  (/home/workspace/epan/wslua/wslua_byte_array.c, ByteArray_base64_decode)   

    RochiaiD  vote  
0   1.0       1.0   
1   1.0       1.0   
2   1.0       1.0   
3   1.0       1.0   
4   1.0       1.0   
5   1.0       1.0   
6   1.0       1.0   
7   1.0       1.0   
8   1.0       1.0   
9   1.0       1.0   
10  1.0       1.0   
11  1.0       1.0   
12  1.0       1.0   
13  1.0       1.0   
14  1.0       1.0

Top 5 methods:


meth  \
0  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)               
1  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)                     
2  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)                 
3  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
4  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)                           

         vote  rank  
0  105.000000  1.0   
1  93.000000   2.0   
2  72.000000   3.0   
3  63.993164   4.0   
4  62.000000   5.0

Top 1:


meth  \
0  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   

    vote  rank  
0  105.0  1.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0    2887  0    1    33   0    1.0   1.0      4981.0  4981.0    
1    2886  0    1    33   0    1.0   1.0      4981.0  4981.0    
2    2884  0    1    33   0    1.0   1.0      4981.0  4981.0    
3    2881  0    1    33   0    1.0   1.0      4981.0  4981.0    
4    2877  0    1    33   0    1.0   1.0      4981.0  4981.0    
..    ... ..   ..    ..  ..    ...   ...         ...     ...    
100  2582  0    1    33   0    1.0   1.0      4981.0  4981.0    
101  2581  0    1    33   0    1.0   1.0      4981.0  4981.0    
102  2580  0    1    33   0    1.0   1.0      4981.0  4981.0    
103  2579  0    1    33   0    1.0   1.0      4981.0  4981.0    
104  2577  0    1    33   0    1.0   1.0      4981.0  4981.0    

                                             file                    func  \
0    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
1    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
2    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
3    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
4    /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
..                                            ...                     ...   
100  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
101  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
102  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
103  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   
104  /home/workspace/epan/dissectors/packet-sip.c  dissect_sip_via_header   

                                                                       meth  \
0    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
1    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
2    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
3    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
4    (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
..                                                                      ...   
100  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
101  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
102  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
103  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   
104  (/home/workspace/epan/dissectors/packet-sip.c, dissect_sip_via_header)   

     RochiaiD  vote  
0    1.0       1.0   
1    1.0       1.0   
2    1.0       1.0   
3    1.0       1.0   
4    1.0       1.0   
..   ...       ...   
100  1.0       1.0   
101  1.0       1.0   
102  1.0       1.0   
103  1.0       1.0   
104  1.0       1.0   

[105 rows x 14 columns]

Top 2:


meth  vote  \
0  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)  93.0   

   rank  
0  2.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   802   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   800   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   797   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   795   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   794   0    1    33   0    1.0   1.0      4981.0  4981.0    
..  ...  ..   ..    ..  ..    ...   ...         ...     ...    
88  538   0    1    33   0    1.0   1.0      4981.0  4981.0    
89  537   0    1    33   0    1.0   1.0      4981.0  4981.0    
90  536   0    1    33   0    1.0   1.0      4981.0  4981.0    
91  535   0    1    33   0    1.0   1.0      4981.0  4981.0    
92  508   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                              file            func  \
0   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
1   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
2   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
3   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
4   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
..                                             ...             ...   
88  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
89  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
90  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
91  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   
92  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_new   

                                                                meth  \
0   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
1   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
2   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
3   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
4   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
..                                                               ...   
88  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
89  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
90  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
91  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   
92  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_new)   

    RochiaiD  vote  
0   1.0       1.0   
1   1.0       1.0   
2   1.0       1.0   
3   1.0       1.0   
4   1.0       1.0   
..  ...       ...   
88  1.0       1.0   
89  1.0       1.0   
90  1.0       1.0   
91  1.0       1.0   
92  1.0       1.0   

[93 rows x 14 columns]

Top 3:


meth  vote  \
0  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)  72.0   

   rank  
0  3.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   938   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   936   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   933   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   931   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   930   0    1    33   0    1.0   1.0      4981.0  4981.0    
..  ...  ..   ..    ..  ..    ...   ...         ...     ...    
67  809   0    1    33   0    1.0   1.0      4981.0  4981.0    
68  808   0    1    33   0    1.0   1.0      4981.0  4981.0    
69  807   0    1    33   0    1.0   1.0      4981.0  4981.0    
70  806   0    1    33   0    1.0   1.0      4981.0  4981.0    
71  805   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                              file                func  \
0   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
1   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
2   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
3   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
4   /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
..                                             ...                 ...   
67  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
68  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
69  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
70  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   
71  /home/workspace/epan/wslua/wslua_proto_field.c  ProtoField_integer   

                                                                    meth  \
0   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
1   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
2   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
3   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
4   (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
..                                                                   ...   
67  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
68  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
69  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
70  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   
71  (/home/workspace/epan/wslua/wslua_proto_field.c, ProtoField_integer)   

    RochiaiD  vote  
0   1.0       1.0   
1   1.0       1.0   
2   1.0       1.0   
3   1.0       1.0   
4   1.0       1.0   
..  ...       ...   
67  1.0       1.0   
68  1.0       1.0   
69  1.0       1.0   
70  1.0       1.0   
71  1.0       1.0   

[72 rows x 14 columns]

Top 4:


meth  \
0  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   

        vote  rank  
0  63.993164  4.0

Voters:


line  e_p  e_f  n_p  n_f      Sop2   Sochiai    Rop2  Rochiai  \
0    571   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
1    570   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
2    569   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
3    568   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
4    567   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
..   ...  ..   ..    ..  ..         ...       ...     ...     ...    
176  335   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
177  330   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
178  328   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
179  327   1    1    32   0    0.970588  0.707107  7651.0  7651.0    
180  325   1    1    32   0    0.970588  0.707107  7651.0  7651.0    

                                       file  \
0    /home/workspace/epan/reassemble_test.c   
1    /home/workspace/epan/reassemble_test.c   
2    /home/workspace/epan/reassemble_test.c   
3    /home/workspace/epan/reassemble_test.c   
4    /home/workspace/epan/reassemble_test.c   
..                                      ...   
176  /home/workspace/epan/reassemble_test.c   
177  /home/workspace/epan/reassemble_test.c   
178  /home/workspace/epan/reassemble_test.c   
179  /home/workspace/epan/reassemble_test.c   
180  /home/workspace/epan/reassemble_test.c   

                                         func  \
0    test_fragment_add_seq_partial_reassembly   
1    test_fragment_add_seq_partial_reassembly   
2    test_fragment_add_seq_partial_reassembly   
3    test_fragment_add_seq_partial_reassembly   
4    test_fragment_add_seq_partial_reassembly   
..                                        ...   
176  test_fragment_add_seq_partial_reassembly   
177  test_fragment_add_seq_partial_reassembly   
178  test_fragment_add_seq_partial_reassembly   
179  test_fragment_add_seq_partial_reassembly   
180  test_fragment_add_seq_partial_reassembly   

                                                                                   meth  \
0    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
1    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
2    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
3    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
4    (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
..                                                                                  ...   
176  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
177  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
178  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
179  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   
180  (/home/workspace/epan/reassemble_test.c, test_fragment_add_seq_partial_reassembly)   

     RochiaiD      vote  
0    2.0       0.353553  
1    2.0       0.353553  
2    2.0       0.353553  
3    2.0       0.353553  
4    2.0       0.353553  
..   ...            ...  
176  2.0       0.353553  
177  2.0       0.353553  
178  2.0       0.353553  
179  2.0       0.353553  
180  2.0       0.353553  

[181 rows x 14 columns]

Top 5:


meth  vote  rank
0  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  62.0  5.0

Voters:


line  e_p  e_f  n_p  n_f  Sop2  Sochiai    Rop2  Rochiai  \
0   532   0    1    33   0    1.0   1.0      4981.0  4981.0    
1   531   0    1    33   0    1.0   1.0      4981.0  4981.0    
2   529   0    1    33   0    1.0   1.0      4981.0  4981.0    
3   527   0    1    33   0    1.0   1.0      4981.0  4981.0    
4   520   0    1    33   0    1.0   1.0      4981.0  4981.0    
..  ...  ..   ..    ..  ..    ...   ...         ...     ...    
57  450   0    1    33   0    1.0   1.0      4981.0  4981.0    
58  449   0    1    33   0    1.0   1.0      4981.0  4981.0    
59  448   0    1    33   0    1.0   1.0      4981.0  4981.0    
60  447   0    1    33   0    1.0   1.0      4981.0  4981.0    
61  441   0    1    33   0    1.0   1.0      4981.0  4981.0    

                                         file           func  \
0   /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
1   /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
2   /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
3   /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
4   /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
..                                        ...            ...   
57  /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
58  /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
59  /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
60  /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   
61  /home/workspace/epan/wslua/wslua_struct.c  Struct_unpack   

                                                          meth  RochiaiD  vote  
0   (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
1   (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
2   (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
3   (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
4   (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
..                                                         ...  ...       ...   
57  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
58  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
59  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
60  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   
61  (/home/workspace/epan/wslua/wslua_struct.c, Struct_unpack)  1.0       1.0   

[62 rows x 14 columns]

# 감상

Max aggregation으로 top이지만 그게 683개. line 수가 많은 편이 아니어서 voting으로도 성적이 좋지 않음